### PR TITLE
add etcd watch design documentation

### PR DESCRIPTION
**Pull request by StupidHod**
Hope for your advice, @xiang90 @AdoHe

### CODE REVIEW

**Comment by AdoHe** [comment]
👍 this should be a good beginning of watch.

File: Documentation/EtcdWatch.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
+# Overview
+
+The EtcdWatch provide methods to watch on a key and cancel a watcher, if the watcher is disconnected on error, it will be resumed automatically.
```
Comment (hongchaodeng) [inline Documentation/EtcdWatch.md]: "cancel a watcher. If ..."
Use "."

File: Documentation/EtcdWatch.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+## requestStream instance
+
+StreamObserver instance
```
Comment (hongchaodeng) [inline Documentation/EtcdWatch.md]: Why do you have "StreamObserver instance" here?
I'm really confused. You have mostly talked about requestStream below?

**Comment by hongchaodeng** [comment]
@StupidHod @AdoHe 
Maybe we can start with the goals to have EtcdWatcher, and then talk about details about how to implement each?

File: Documentation/EtcdWatch.md (hunk 2 above)
Comment (StupidHod) [inline Documentation/EtcdWatch.md]: It returned by gRPC call watch.

**Comment by AdoHe** [comment]
@hongchaodeng yes, we should examine the goals before we start the implementation.

**Comment by xiang90** [comment]
@StupidHod So this works pretty similar to the go version?

File: Documentation/EtcdWatch.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
+# Overview
+
+The EtcdWatch provide methods to watch on a key and cancel a watcher. If the watcher is disconnected on error, it will be resumed automatically.
```
Comment (heyitsanthony) [inline Documentation/EtcdWatch.md]: s/key/key interval

File: Documentation/EtcdWatch.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+# Goals
+
+1. The watch client should process watch request, the watch client create watch request for client with key, option(revision, delete only, put only and end key) and register callback for watch request.
+
+2. Notification, when watch client receive event from etcd server, it will should call registered callback.
+
+3. Resume, when the watch client disconnect with etcd server, the etcd client should automatically resume all the watch requests and get notification from the latest received revision.
```
Comment (heyitsanthony) [inline Documentation/EtcdWatch.md]: The go client doesn't send any notification on resume.

File: Documentation/EtcdWatch.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+4. Cancel watch request, the etcd client should process watch cancellation and filter all the notification after cancellation request.
+
+# Implementation
+
+The Etcd client process watch request with [watch function](#watch-function), process notification with [processEvents function](#processevents-function) , process resume with [resume function](#resume-function) and process cancel with [cancelWatch function](#cancelwatch-function).
+
+## watch function
+
+Watch watches on a key or prefix.
```
Comment (heyitsanthony) [inline Documentation/EtcdWatch.md]: s/key or prefix/key interval

File: Documentation/EtcdWatch.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+1. Remove the respond watcher from [cancelWatchers](#cancelwatchers).
+2. call `onCancel` callback.
+
+
+## resumeWatchers function
+
+Resume all the the watchers on new requestStream.
+
+1. Build new watch creation request for old watcher with last revision.
```
Comment (heyitsanthony) [inline Documentation/EtcdWatch.md]: I think this should be last revision + 1 since watch sends events that are >= watch revision.

**Comment by xiang90** [comment]
The design doc looks good. Let's clean it up and get it merged.

File: Documentation/EtcdWatch.md (hunk 4 above)
Comment (StupidHod) [inline Documentation/EtcdWatch.md]: @heyitsanthony Sorry, I use uncorrect word, it should be ' get events from the latest received revision', set the new watch option with revision = last_revision + 1.

File: Documentation/EtcdWatch.md (hunk 6 above)
Comment (StupidHod) [inline Documentation/EtcdWatch.md]: @heyitsanthony , I will set the revision = last revision, and filter the events with last revision to clear the resuming status.

File: Documentation/EtcdWatch.md (hunk 5 above)
Comment (StupidHod) [inline Documentation/EtcdWatch.md]: Ok, got it, tks.

**Comment by StupidHod** [comment]
@xiang90 , thanks for your advice, I have cleared it up with your tips.

File: Documentation/EtcdWatch.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
+# Overview
```
Comment (xiang90) [inline Documentation/EtcdWatch.md]: replace all Etcd with etcd (including file name)

Comment (StupidHod) [inline Documentation/EtcdWatch.md]: OK.

**Comment by hongchaodeng** [comment]
LGTM
Let's move forward with implementation.
Thanks @StupidHod !

**Pull request merged**